### PR TITLE
 'num' support a tensor as input in tf.unstack(#17266)

### PR DESCRIPTION
**Pull request by jason9693**

### CODE REVIEW

File: tensorflow/python/ops/array_ops.py [hunk 1]
```diff
@@ -1021,6 +1021,14 @@ def unstack(value, num=None, axis=0, name="unstack"):
       num = value_shape[axis].value
   if num is None:
     raise ValueError("Cannot infer num from shape %s" % value_shape)
+  elif isinstance(num, ops.Tensor):
+    if num.shape.ndims == 0:
```
Comment (rmlarsen) [inline tensorflow/python/ops/array_ops.py]: The shape may not be known at graph construction time. If the rank is unknown statically, use tf.rank and tf.assert to insert a runtime check in the graph.

Comment (rmlarsen) [reply tensorflow/python/ops/array_ops.py]: Please add unit tests that test this with both static and dynamic shapes.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/array_ops.py (hunk 1 above)
Comment (jason9693) [reply tensorflow/python/ops/array_ops.py]: @rmlarsen  the shape you said is in 'num'. right??

Comment (jason9693) [reply tensorflow/python/ops/array_ops.py]: @rmlarsen  I fixed.

File: tensorflow/python/ops/array_ops.py [hunk 2]
```diff
@@ -1021,6 +1022,13 @@ def unstack(value, num=None, axis=0, name="unstack"):
       num = value_shape[axis].value
   if num is None:
     raise ValueError("Cannot infer num from shape %s" % value_shape)
+  elif isinstance(num, ops.Tensor):
+    if gen_math_ops.equal(rank(num),0):
```
Comment (rmlarsen) [inline tensorflow/python/ops/array_ops.py]: space after comma

File: tensorflow/python/kernel_tests/unstack_op_test.py [hunk 3]
```diff
@@ -161,6 +161,17 @@ def testZeroLengthDim(self):
       y = array_ops.unstack(x, axis=1)[0].eval()
       self.assertEqual(y.shape, (0, 2))
 
+  def testNumWithTensor(self):
+    dynamic_feed_shape = (2,None)
+    static_feed_shape_dict = {'scalar': (2, ()),'one_dim': (2, [1])}
+    x = array_ops.constant([[1,2,3],[4,5,6]])
+    with self.test_session() as sess:
+      for feed, shape in (dynamic_feed_shape, static_feed_shape_dict['scalar'],static_feed_shape_dict['one_dim']):
```
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/unstack_op_test.py]: Can you run pyformat or make sure the pylint is happy?

**Review by rmlarsen — DISMISSED** [review]

File: tensorflow/python/kernel_tests/unstack_op_test.py (hunk 3 above)
Comment (jason9693) [reply tensorflow/python/kernel_tests/unstack_op_test.py]: I forgot to check pylint.
now, I checked and fixed

**Review by rmlarsen — DISMISSED** [review]

**Comment by jason9693** [comment]
@rmlarsen I think We should modify gen_array_ops.py, But this file is MACHINE GENERATED. So, we should edit array_ops.cc. However, I don't know how to generate gen_array_ops.py file by array_ops.cc. Can you give me a way to generate this file??

**Comment by rmlarsen** [comment]
@jason9693 gen_array_ops.py gets generate automatically during the build.

**Comment by jason9693** [comment]
@rmlarsen  Is it only available to build in ubuntu or windows os, not in mac?

**Comment by rmlarsen** [comment]
@jason9693 It is available on all platforms.

**Comment by drpngx** [comment]
@jason9693 there are many errors, including the sanity check. Could you fix and push again?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 44 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Pull request closed without merging**